GET pagination #141

tylerelston opened the pull request:
Implemented pagination and wrote tests for:
- authors
- posts
- comments

resolves #141

JawadHossain commented on code/api/tests/test_comments.py:
```diff
@@ -314,6 +313,111 @@ def test_multi_comments_same_author(self):
+
         expected_len = len(str(expected))
         res_data_len = len(str(res_data))
         self.assertEqual(expected_len, res_data_len,
```
This test fails, as the api returns comments with id 1 and 3. Instead of 1 and 2

JawadHossain commented on code/api/utility.py:
```diff
@@ -0,0 +1,23 @@
+        if page < 1 or size < 1:
+            return HttpResponseBadRequest("Malformed query: page and size must be > 0")
+    except:
+        print('hi')
```
print

JawadHossain reviewed (changes requested):
Nice! 
Does Django support pagination queries? If so would be nice to only query the required data from the database.
A test to check content on the second page would be great.
A test failing can you please check?

tylerelston commented:
> Nice! Does Django support pagination queries? If so would be nice to only query the required data from the database. A test to check content on the second page would be great. A test failing can you please check?

Refactored the pagination function to use django's built in Paginator and added multiple page tests.
Manually testing the queries works fine, but for some reason the comment tests are inconsistent and sometimes pass, sometimes fail. Still trying to figure out why

tylerelston commented:
Fixed the test, should be good now

JawadHossain reviewed (approved):
Thanks for the fix!

debecerra commented on code/api/utility.py:
```diff
@@ -0,0 +1,12 @@
+            return HttpResponseBadRequest("Malformed query: page and size must be > 0")
+    except:
+        return HttpResponseBadRequest("Malformed query")
+    p = Paginator(data, size)
+    return p.page(page)
```
Server crashes and returns a 500 status code if we request a page with no results
`django.core.paginator.EmptyPage: That page contains no results`

Can we add a simple try-except block to catch this error? Just so it doesn't return a 5XX code. I think it would make sense to just return an empty list but anything other than a server crash is probably good.

tylerelston replied:
Sounds good

debecerra commented on code/api/views.py:
```diff
@@ -46,13 +47,20 @@ class AuthorsView(View):
+            size = int(size)
+            authors = getPaginated(authors, page, size)
 
-        authors = [author.as_json() for author in LocalAuthor.objects.all()]
-
+        authors = [author.as_json() for author in authors]
```
`getPaginated` returns an `HttpResponseBadRequest` object if page < 1 or size < 1, but then this get's assigned to `author`. Then the code crashes on `author.as_json()`

tylerelston replied:
Moved error handling to views.py

debecerra reviewed (changes requested):
Looks great! 

Two comments about server crash on a few edge cases I was trying out.
